ide/posts/2020-06-50-ghcide-and-nixpkgs.md

```diff
@@ -0,0 +1,321 @@
+   packages not on stackage.
+
+Now solution two is in some sense less complex to use, because at compile time
+you don‘t need to construct a build plan. Of course, as I said, today cabal can
+do this for you very smoothly, which is why I personally prefer the first
+   approach.
+
+### Haskell in nixpkgs - pkgs.hackagePackages
+
+But how does nixpkgs do it? Well basically solution two. Everyday a cronjob
+pulls a list of **all packages from a pinned stack LTS release** and creates a
+derivation for every one of them. It also pulls **all other packages from
+hackage** and creates a derivation for the **latest released version** of them.
+(This happens on the haskell-updates branch of nixpkgs which get‘s normally
+merged into nixpkgs master i.e. unstable once per week.) So then, you ask, how
+does cabal2nix do dependency resolution? Well the short form is, it doesn‘t.
+What I mean by that is: It completely ignores any version bounds given in a
+cabal file or a pinned stack LTS release. It will just **take the one version
+of every dependency that is present in nixpkgs** by the method I told you
+above.
+
+When I first learned about this I thought this was ludicrous. How can someone
+device a not-solution like this? This is prone to fail. And indeed it does. A
+large number of packages just don‘t build this way. Concretely the build will
+either fail at compile time or more often cabal will complain that it can‘t
+create a build plan. What that actually means: cabal says the one build plan we
+provided it with is invalid because it does not match the given version bounds.
+duh. So that packages get automatically marked broken after hydra, the nixos
+build server, fails to build them. And oh boy, there are a lot of Haskell
+packages broken in nixpkgs.
+
+This is obviously very frustrating and so I vented my frustration in the nixos
+discourse. I clearly hadn‘t understood a lot of what I just explained at that
+time, so in retrospect definitely not one of my proudest moments.
+
+Now what would be a suitable alternative to this for nixpkgs? Tough to say. We
+could try to use some solution like the go, rust or node ecosystem and check in
+a build plan for every package. Actually that can be a nice solution and if you
+are interested in that you should definitely checkout the `haskell.nix`
+infrastructure.
+But that really does not go well together with providing all of hackage in
+nixpkgs. For starters having every version of every Haskell package in nixpkgs
+would already be very verbose. And it would lead to tons of different builds
+for the same package even with the same version, one for every different valid
+build plan of dependencies. And still if nixpkgs would only build the build
+plans necessary to build every package once, your project would probably use
+a different build plan and therefore you would need to compile a lot of
+dependencies by yourself and couldn‘t use the precompiled derivations from
+nixpkgs.
```

Comment:
Is this really true? Surely you would only need to compile each version of each package once and then could use
a combination of these dependencies depending on your build plan?

Reply:
Yes, this is true, I think. If we compile let‘s say lens-2.0 on hydra, against text-3.0 and your package relies on lens-2.0 but uses text-4.0, you can‘t use the precompiled version of lens-2.0 even though it is the exact version that you want.

Reply:
I was also confused by this paragraph. In reality what nixpkgs does with *everything* is not conceptually different from what Stackage does, so it's really not weird at all.

Reply:
I will have a look at that paragraph to see if I can clear it up.

Reply:
Gave it a shot.